MongoDb with Panache: Add multi-tenancy support to MongoDb Panache throught dynamic database selection

pedroh-pereira opened the pull request:
Based on #14789

It consists in a interface called MongoDatabaseResolver that can be implemented to get the database name to be used on mongo operations execution
So if I have a MongoDatabaseResolver implementation the database to be used in the current mongo operation will be the return of resolve() method, but the resolver will be successful only for the entities that doesn't specify the database through the @MongoEntity(database="")

Disclaimer:
After this PR be approved I can help to provide a detailed documentation
My other PR #29192 was closed accidentaly

@loicmathieu @evanchooly @geoand

pedroh-pereira committed:
chore: clean up MongoOperations and ReactiveMongoOperations (mongo-panache)

loicmathieu commented on integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/multitenancy/CustomMongoDatabaseResolver.java:
```diff
@@ -0,0 +1,25 @@
+import io.quarkus.mongodb.panache.common.MongoDatabaseResolver;
+import io.vertx.ext.web.RoutingContext;
+
+@Unremovable
```
Should not be needed, you can make all beans unremovable from the deployment module.

pedroh-pereira replied:
May be it sounds a dumb question, Are you saying the deployment module from mongo-panache or integration test?

loicmathieu commented on integration-tests/mongodb-panache/src/main/resources/application.properties:
```diff
@@ -8,3 +8,7 @@ quarkus.mongodb.cl2.write-concern.journal=false
 #quarkus.log.category."io.quarkus.mongodb.panache.runtime".level=DEBUG
 quarkus.mongodb.metrics.enabled=true
+
+# avoid CustomMongoDatabaseResolver initialization to used profiles different than MongoDatabaseResolverProfile
+%test.quarkus.arc.exclude-types=io.quarkus.it.mongodb.panache.multitenancy.CustomMongoDatabaseResolver
+%prod.quarkus.arc.exclude-types=io.quarkus.it.mongodb.panache.multitenancy.CustomMongoDatabaseResolver
```
It's not very easy to understand why it's needed, maybe a better way would be to move the test to the deployment module where you can control the beans that are inside a test by using `@QuarkusUnitTest`.

loicmathieu reviewed (changes requested):
Hi,
Sorry for the delay, the implementation seems OK but the test is a bit complicated and I would prefer having it inside the deployment module with a minimal test instead of testing multiple REST operations.
In the deployment module you can also easily test things in isolation using `@QuarkusUnitTest`, this avoid using profiles.

pedroh-pereira commented:
@loicmathieu when you say deployment module, do you mean the deployment  module from extension itself or integration tests?

loicmathieu commented:
@pedroh-pereira the deployment module from the extension, in it you can use `QuarkusUnitTest` to cherrypick classes and config files you want to include for the test. See https://github.com/quarkusio/quarkus/blob/main/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/DefaultMongoClientConfigTest.java for example.

pedroh-pereira commented:
> @pedroh-pereira the deployment module from the extension, in it you can use `QuarkusUnitTest` to cherrypick classes and config files you want to include for the test. See https://github.com/quarkusio/quarkus/blob/main/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/DefaultMongoClientConfigTest.java for example.

Thank you, done

pedroh-pereira commented:
> @pedroh-pereira the deployment module from the extension, in it you can use `QuarkusUnitTest` to cherrypick classes and config files you want to include for the test. See https://github.com/quarkusio/quarkus/blob/main/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/DefaultMongoClientConfigTest.java for example.

@loicmathieu Thanks, I've tried to follow your suggestions

pedroh-pereira commented:
I think it's ready to be reviewed again.
@loicmathieu @evanchooly @gsmet @geoand

loicmathieu commented on docs/src/main/asciidoc/mongodb-panache.adoc:
```diff
@@ -1238,3 +1238,285 @@ by the presence of the marker file `META-INF/panache-archive.marker`. Panache in
+----
+
+
+=== Configuring the application
+
+The same mongo connection will be used for all tenants, so a database has to be created for every tenant.
+
+[source,properties]
+----
+quarkus.mongodb.connection-string=mongodb://login:pass@mongo:27017
+# The default database
+quarkus.mongodb.database=sanjoka
+----
```
Is it relly needed ? As the same connection will be used and the database config property will not be used I think this can be removed.

pedroh-pereira replied:
if the resolver returns null, the default database will be used:
![image](https://user-images.githubusercontent.com/47221187/210802286-01ecead8-0f77-4d51-af96-e8cca5f18215.png)

loicmathieu replied:
Yes, so when using multitenancy this is optional in case we didn't find the tenant.

loicmathieu commented on docs/src/main/asciidoc/mongodb-panache.adoc:
```diff
@@ -1238,3 +1238,285 @@ by the presence of the marker file `META-INF/panache-archive.marker`. Panache in
+import io.restassured.specification.RequestSpecification;
+
+@QuarkusTest
+@QuarkusTestResource(MongoReplicaSetTestResource.class)
+@DisabledOnOs(OS.WINDOWS)
```
Please remove them, users will use Dev Services

pedroh-pereira replied:
Done

loicmathieu commented on docs/src/main/asciidoc/mongodb-panache.adoc:
```diff
@@ -1238,3 +1238,285 @@ by the presence of the marker file `META-INF/panache-archive.marker`. Panache in
+        String endpoint = "/persons";
+
+        RestAssured.defaultParser = Parser.JSON;
+        RestAssured.config
+                .objectMapperConfig(new ObjectMapperConfig().jackson2ObjectMapperFactory((type, s) -> new ObjectMapper()
+                        .registerModule(new Jdk8Module())
+                        .registerModule(new JavaTimeModule())
+                        .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+                        .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)));
```
Is it really needed ?

pedroh-pereira replied:
It's not needed, I fixed it. Thanks!

pedroh-pereira commented on docs/src/main/asciidoc/mongodb-panache.adoc:
```diff
@@ -1238,3 +1238,285 @@ by the presence of the marker file `META-INF/panache-archive.marker`. Panache in
+    
+}
+---- 
+====
```
You're right, I already removed it.

pedroh-pereira commented on docs/src/main/asciidoc/mongodb-panache.adoc:
```diff
@@ -1238,3 +1238,285 @@ by the presence of the marker file `META-INF/panache-archive.marker`. Panache in
+
+[IMPORTANT]
+====
+The database selection priority order is followed by: `@MongoEntity(database="mizain")`, `MongoDatabaseResolver` 
```
Done

pedroh-pereira commented on docs/src/main/asciidoc/mongodb-panache.adoc:
```diff
@@ -1238,3 +1238,285 @@ by the presence of the marker file `META-INF/panache-archive.marker`. Panache in
+}
+----
+<1> Annotate the MongoDatabaseResolver implementation with the `@Unremovable` qualifier to tell Quarkus it must not 
+be auto removed, because behind the scenes it was not being injected by using `@Inject`.
```
Done

pedroh-pereira commented on docs/src/main/asciidoc/mongodb-panache.adoc:
```diff
@@ -1238,3 +1238,285 @@ by the presence of the marker file `META-INF/panache-archive.marker`. Panache in
+=== Writing the application
+
+In order to resolve the tenant from incoming requests and map it to a specific database, you must create an implementation 
+for `io.quarkus.mongodb.panache.common.MongoDatabaseResolver` interface.
```
Done

pedroh-pereira commented on docs/src/main/asciidoc/mongodb-panache.adoc:
```diff
@@ -141,8 +141,8 @@ The `quarkus.mongodb.database` property will be used by MongoDB with Panache to
-* the name of the client for multi-tenant application, see xref:mongodb.adoc#multiple-mongodb-clients[Multiple MongoDB Clients]. Otherwise, the default client will be used.
-* the name of the database, otherwise, the `quarkus.mongodb.database` property will be used.
+* the name of the client for multitenant application, see xref:mongodb.adoc#multiple-mongodb-clients[Multiple MongoDB Clients]. Otherwise, the default client will be used.
+* the name of the database, otherwise the `quarkus.mongodb.database` property or an link:{mongodb-doc-root-url}#multitenancy[`MongoDatabaseResolver`] implementation will be used.
```
Done

pedroh-pereira commented on docs/src/main/asciidoc/mongodb-panache.adoc:
```diff
@@ -1238,3 +1238,285 @@ by the presence of the marker file `META-INF/panache-archive.marker`. Panache in
+    }
+}
+----
+
```
Sorry, I didn't understand

pedroh-pereira commented:
Ready for review again!
@loicmathieu @evanchooly @geoand @gsmet

loicmathieu commented on docs/src/main/asciidoc/mongodb-panache.adoc:
```diff
@@ -1238,3 +1238,218 @@ by the presence of the marker file `META-INF/panache-archive.marker`. Panache in
+
+[IMPORTANT]
+====
+The database selection priority order is as follow: `@MongoEntity(database="mizain")`, `MongoDatabaseResolver` 
```
```suggestion
The database selection priority order is as follow: `@MongoEntity(database="mizain")`, `MongoDatabaseResolver`, 
```

pedroh-pereira replied:
Done

loicmathieu commented on extensions/panache/mongodb-panache-common/deployment/src/test/java/io/quarkus/mongodb/panache/common/MongoDatabaseResolverTest.java:
```diff
@@ -0,0 +1,199 @@
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addAsResource(new StringAsset(
+                            "quarkus.arc.remove-unused-beans=false\n" +
```
Can you remove the `quarkus.arc.remove-unused-beans=false` and check it stills work ?

pedroh-pereira replied:
I already had tried before and it doesn't work, because the BuildStep method responsible to make `MongoDatabaseResolver` unremovable is described in `BasePanacheMongoResourceProcessor` that is an abstract class and it implementation is in another project (mongodb-panache)

loicmathieu replied:
OK, make thanks

loicmathieu reviewed (changes requested):
Overall looks good to me, can you process with the few little changes I ask so we can merge this one ?
And again thanks for your contribution, I know it's been a long process but hopefully it'll be merged soon ;)

pedroh-pereira commented:
> Overall looks good to me, can you process with the few little changes I ask so we can merge this one ? And again thanks for your contribution, I know it's been a long process but hopefully it'll be merged soon ;)

 Don't worry about that, I completely understand that this long process is required for an assertive delivery and thanks for the review!

pedroh-pereira committed:
feat: MongoDb with Panache - Add dynamic database selection support

It supports multi-tenancy implementation following database per tenant approach, It consists in a interface called MongoDatabaseResolver that can be implemented to get the database name to be used on mongo operations execution. So if I have a MongoDatabaseResolver implementation the database to be used in the current mongo operation will be the return of resolve() method, but the resolver will be successful only for the entities that doesn't specify the database through the @MongoEntity(database="")

Example:
There is a "CustomMongoDatabaseResolver", a insert operation is executing, during mongo database selection the resolve method will be called and if it have returned "customerA", the selected mongo database used to execute this operation was "customerA"

loicmathieu reviewed (approved):
LGTM, thanks
